file_store/src/traits/report_id.rs

```diff
@@ -0,0 +1,47 @@
+use crate::{
+    lora_beacon_report::{LoraBeaconIngestReport, LoraBeaconReport},
+    lora_witness_report::{LoraWitnessIngestReport, LoraWitnessReport},
+};
+use chrono::{DateTime, Utc};
+use sha2::{Digest, Sha256};
+
+pub trait IngestId {
+    fn ingest_id(&self) -> Vec<u8>;
+}
+
+pub trait ReportId {
+    fn report_id(&self, received_ts: DateTime<Utc>) -> Vec<u8>;
+}
+
+macro_rules! impl_ingest_id {
+    ($report_type:ty) => {
+        impl IngestId for $report_type {
+            fn ingest_id(&self) -> Vec<u8> {
+                let mut id: Vec<u8> = self.report.data.clone();
+                let mut public_key = self.report.pub_key.to_vec();
+                id.append(&mut self.received_timestamp.to_string().as_bytes().to_vec());
+                id.append(&mut public_key);
+                Sha256::digest(&id).to_vec()
+            }
+        }
+    };
+}
+
+macro_rules! impl_report_id {
+    ($report_type:ty) => {
+        impl ReportId for $report_type {
+            fn report_id(&self, received_ts: DateTime<Utc>) -> Vec<u8> {
+                let mut id: Vec<u8> = self.data.clone();
+                let mut public_key = self.pub_key.to_vec();
+                id.append(&mut received_ts.to_string().as_bytes().to_vec());
+                id.append(&mut public_key);
+                Sha256::digest(&id).to_vec()
+            }
+        }
+    };
+}
```

Comment:
i'm not sure about the utility of these macros since they depend on the fields of the implementing structs pretty heavily; from what i can tell, it's more idiomatic to just put the `impl ReportId for ReportType` block, typically in the same file that defines the `ReportType` struct (whichever that might be)

Reply:
well i mean thats the way it was previously.  this change was a patch created by vihu to move the ID generation out of the report `impl` and into their own trait.

Reply:
it'd be great to have the signature in here as well